src/darwin/Framework/CHIP/MTRDevice.mm

```diff
@@ -284,122 +348,125 @@ - (void)_handleEventReport:(NSArray<NSDictionary<NSString *, id> *> *)eventRepor
+                                  std::unique_ptr<ClusterStateCache> clusterStateCache;
+                                  callback = std::make_unique<SubscriptionCallback>(
+                                      ^(NSArray * value) {
+                                          MTR_LOG_INFO("%@ got attribute report %@", self, value);
+                                          dispatch_async(self.queue, ^{
+                                              // OnAttributeData (after OnReportEnd)
+                                              [self _handleAttributeReport:value];
+                                          });
+                                      },
+                                      ^(NSArray * value) {
+                                          MTR_LOG_INFO("%@ got event report %@", self, value);
+                                          dispatch_async(self.queue, ^{
+                                              // OnEventReport (after OnReportEnd)
+                                              [self _handleEventReport:value];
+                                          });
+                                      },
+                                      ^(NSError * error) {
+                                          MTR_LOG_INFO("%@ got subscription error %@", self, error);
+                                          dispatch_async(self.queue, ^{
+                                              // OnError
+                                              [self _handleSubscriptionError:error];
+                                          });
+                                      },
+                                      ^(NSError * error, NSNumber * resubscriptionDelay) {
+                                          MTR_LOG_INFO("%@ got resubscription error %@ delay %@", self, error, resubscriptionDelay);
+                                          dispatch_async(self.queue, ^{
+                                              // OnResubscriptionNeeded
+                                              [self _handleResubscriptionNeeded];
+                                          });
+                                      },
+                                      ^(void) {
+                                          MTR_LOG_INFO("%@ got subscription established", self);
+                                          dispatch_async(self.queue, ^{
+                                              // OnSubscriptionEstablished
+                                              [self _handleSubscriptionEstablished];
+                                          });
+                                      },
+                                      ^(void) {
+                                          MTR_LOG_INFO("%@ got subscription done", self);
+                                          dispatch_async(self.queue, ^{
+                                              // OnDone
+                                              [self _handleSubscriptionReset];
```

Comment:
Do we really want to schedule a resubscribe in OnDone?  I guess we are relying on the "If this is being triggered from shutdown then invalidate would have been called already and we won't try to resubscribe because we will not have a delegate" bit?

Reply:
OK, sounds good.